ukat/mapping/phase_contrast.py

```diff
@@ -0,0 +1,221 @@
+    min_velocity_phase : list
+        List containing the minimum velocity values (cm/s) per phase.
+    mean_velocity_phase : list
+        List containing the average velocity values (cm/s) per phase.
+    max_velocity_phase : list
+        List containing the maximum velocity values (cm/s) per phase.
+    peak_velocity_phase : list
+        List containing the maximum velocity values (cm/s) per phase.
+    std_velocity_phase : list
+        List containing the std dev of the velocity values (cm/s) per phase.
+    RBF : list
+        List containing the Renal Blood Flow values (ml/min) per phase.
+    stats_table : dictionary
+        A dictionary containing all class attributes that are a list.
+    mean_velocity : float
+        Average velocity (cm/s) accross the different phases.
+    mean_RBF : float
+        Average Renal Blood Flow (ml/min) accross the different phases.
+    resistive_index : float
+        A prognostic marker in renal vascular diseases which range is [0, 1].
+    """
+
+    def __init__(self, velocity_array, affine, mask=None):
+        """Initialise a PhaseContrast class instance.
+
+        Parameters
+        ----------
+        velocity_array : np.ndarray
+            A 3D array containing the velocity images of the phase contrast
+            sequence, i.e. the dimensions of the array are [x, y, p], where p
+            corresponds to the phase (or trigger delay).
+        affine : np.ndarray
+            A matrix giving the relationship between voxel coordinates and
+            world coordinates.
+        mask : np.ndarray, optional
+            A boolean mask of the voxels to fit. Should be the shape of the
+            the raw data.
+        """
+        self.shape = velocity_array.shape
+        self.affine = affine
+        self.pixel_spacing = (np.linalg.norm(self.affine[:3, 1]),
+                              np.linalg.norm(self.affine[:3, 0]))
+        # Generate a mask if there isn't one specified
+        if mask is None:
+            self.mask = np.ones(self.shape, dtype=bool)
+        else:
+            self.mask = np.where(mask == 0, np.nan, mask)
+            # The purpose is for the np.nanmean to work properly later
+            # without including the 0s of the mask
+        self.velocity_array = np.abs(velocity_array * self.mask)
+        self.num_pixels_phase = []
+        self.area_phase = []
+        self.min_velocity_phase = []
+        self.mean_velocity_phase = []
+        self.max_velocity_phase = []
+        self.peak_velocity_phase = []
+        self.std_velocity_phase = []
+        self.RBF = []
+        self.mean_velocity = 0
+        self.mean_RBF = 0
+        self.resistive_index = 0
+        if len(self.shape) == 3:
+            for phase in range(self.shape[-1]):
+                phase_array = self.velocity_array[..., phase]
+                num_pixels = np.count_nonzero(~np.isnan(phase_array))
+                area = (num_pixels * 0.1 * self.pixel_spacing[0] *
```

Comment:
Can this line be moved outside the for loop given it doesn't change on each iteration?

Reply:
I'm not sure that would work in this case because `num_pixels` can be different in each phase. The number of masked pixels isn't necessarily the same for all phases.

Reply:
Yes, `num_pixels` needs to be in the loop (the way you've currently written it, in reality, it could be done more efficiently by doing `x = np.count_nonzeros(~np.isnan(self.velocity_array), axis=(0, 1))` where `x` is the list youre building up in the loop, currently `num_pixels_phase`, but see my other comment about the name of that variable), but `area` is just `num_pixels` times a constant so can be below the loop.